biz.aQute.bndlib/src/aQute/bnd/classfile/ConstantPool.java

```diff
@@ -888,4 +898,316 @@ public void write(DataOutput out) throws IOException {
+			else if (entry instanceof MethodHandleInfo)
+				visitor.visit(i, (MethodHandleInfo) entry);
+			else if (entry instanceof MethodTypeInfo)
+				visitor.visit(i, (MethodTypeInfo) entry);
+			else if (entry instanceof ModuleInfo)
+				visitor.visit(i, (ModuleInfo) entry);
+			else if (entry instanceof NameAndTypeInfo)
+				visitor.visit(i, (NameAndTypeInfo) entry);
+			else if (entry instanceof PackageInfo)
+				visitor.visit(i, (PackageInfo) entry);
+			else if (entry instanceof StringInfo)
+				visitor.visit(i, (StringInfo) entry);
+			else if (entry instanceof String)
+				visitor.visit(i, (String) entry);
+			else if (entry instanceof Number)
+				visitor.visit(i, (Number) entry);
+			else
+				visitor.unknown(i, entry);
+		}
+	}
+
+	/**
+	 * An attribute visitor to safe 'instanceof' avalanches. Each method can
+	 * return a replacement attribute or null. If not null, the attribute is
+	 * replaced. If the {@link AttributeVisitor#REMOVE} is returned, the
+	 * attribute is removed from the list.
+	 */
+
+	@ProviderType
+	public interface AttributeVisitor {
+		/**
+		 * Return if you want this attribute to be removed.
+		 */
+		Attribute REMOVE = new UnrecognizedAttribute("remove", null);
+
+		default Attribute visit(AnnotationDefaultAttribute attr) {
+			return null;
+		}
+
+		default Attribute visit(AnnotationsAttribute attr) {
+			return null;
+		}
+
+		default Attribute visit(BootstrapMethodsAttribute attr) {
+			return null;
+		}
+
+		default Attribute visit(CodeAttribute attr) {
+			return null;
+		}
+
+		default Attribute visit(ConstantValueAttribute attr) {
+			return null;
+		}
+
+		default Attribute visit(DeprecatedAttribute attr) {
+			return null;
+		}
+
+		default Attribute visit(EnclosingMethodAttribute attr) {
+			return null;
+		}
+
+		default Attribute visit(ExceptionsAttribute attr) {
+			return null;
+		}
+
+		default Attribute visit(InnerClassesAttribute attr) {
+			return null;
+		}
+
+		default Attribute visit(LineNumberTableAttribute attr) {
+			return null;
+		}
+
+		default Attribute visit(LocalVariableTableAttribute attr) {
+			return null;
+		}
+
+		default Attribute visit(LocalVariableTypeTableAttribute attr) {
+			return null;
+		}
+
+		default Attribute visit(MethodParametersAttribute attr) {
+			return null;
+		}
+
+		default Attribute visit(ModuleAttribute attr) {
+			return null;
+		}
+
+		default Attribute visit(ModuleMainClassAttribute attr) {
+			return null;
+		}
+
+		default Attribute visit(ModulePackagesAttribute attr) {
+			return null;
+		}
+
+		default Attribute visit(NestHostAttribute attr) {
+			return null;
+		}
+
+		default Attribute visit(NestMembersAttribute attr) {
+			return null;
+		}
+
+		default Attribute visit(RuntimeInvisibleParameterAnnotationsAttribute attr) {
+			return null;
+		}
+
+		default Attribute visit(RuntimeVisibleParameterAnnotationsAttribute attr) {
+			return null;
+		}
+
+		default Attribute visit(PermittedSubclassesAttribute attr) {
```

Comment:
We cannot have preview package type exposed in the classfile package API. The classfile API is currently safe in that the API never references the preview package. We do not want to have to make a major change because a preview package type changes before going final in the JVMS.